### PR TITLE
WIP: Add PostCSS's SugarSS recognition by default

### PR DESCRIPTION
**Pull request by DoumanAsh**
Personally I find [SugarSS](https://github.com/postcss/sugarss) is pretty nice simplified version of CSS syntax.

Since its addition is configured by setting corresponding parse in PostCSS's config, I wonder if it would be ok to add its extension by default?

### CODE REVIEW

**Comment by Kompwu** [comment]
@DoumanAsh Can you help me ?
I'm trying to get `.sss` bundled...

**Comment by DoumanAsh** [comment]
@Kompwu Well as far as I understand currently you'd need to register this extension to be recognized as CSS asset:

```js
const Bundler = require('parcel-bundler');

let bundler = new Bundler('input.js');
bundler.addAssetType('.sss', './assets/CSSAsset');
```

If I understand this API correctly(I'm currently decided to use plain CSS with nesting support

**Comment by Kompwu** [comment]
@DoumanAsh Thanks,
How can i make the same thing without js ?

**Comment by DoumanAsh** [comment]
@Kompwu Sorry I'm not sure as I'm new to parcel myself

**Comment by Kompwu** [comment]
@DeMoorJasper Can you merge it pls ?

**Comment by DeMoorJasper** [comment]
It would be nice if the sugarSS asset actually installed and configured postCSS for the user, this way it works `zero-config`, you can probably create an asset that runs sss through postCSS and then passes it to css. (Hope this makes any sense, lemme know if you need any help)

**Comment by DoumanAsh** [comment]
@DeMoorJasper Ok, I'll try to approach with this suggestion. But what should we do if user already has config?

**Comment by DeMoorJasper** [comment]
@DoumanAsh You can probably load PostCSS and check for SugarSS and just forward it to CSS if it's already installed and configured I guess?

**Comment by Kompwu** [comment]
@DoumanAsh How's going ?

**Comment by DoumanAsh** [comment]
I tried to give it a go by extending existing CSS asset, but it makes me think that it is rather difficult to do.

The problem is not only to load postcss config, but there is also `parse` method that requires you to specify syntax so I'd need to load sugarss in two places it seems

**Comment by Kompwu** [comment]
@DoumanAsh Do what you can and we'll help you with everything else.

**Comment by Kompwu** [comment]
@DoumanAsh How's going?

**Comment by DoumanAsh** [comment]
@Kompwu Sorry, I've been a bit busy lately, I'll try to resume working on it as soon as I have time

**Comment by DoumanAsh** [comment]
Currently I have issue with PostCSS API usage as it seems to cause errors when I attempt to parse `.sss`:
```CssSyntaxError: Missed semicolon````

Though I did set `let root = postcss.parse(code, {from: this.name, to: this.name, syntax: sugarss})` corresponding parser.

I was thinking that I could parse SugarSSS into CSS and then rely on existing `CSSAsset` but it seems I misunderstood the API as it ignores provided by me syntax.

**Comment by Kompwu** [comment]
@DoumanAsh That's weird... i have no idea mate.

**Comment by Kompwu** [comment]
@DoumanAsh Any news ?
@devongovett Hey, can you help us ?

**Comment by DoumanAsh** [comment]
@Kompwu Sorry, recently I'm busy with other things and didn't look into PostCSS

**Comment by DoumanAsh** [comment]
@Kompwu Sorry I'm not actively using parcel as of the moment and kinda busy so I'm not planning to resume work on it until need arise again.

I don't mind someone to carry on, or leave issue open so that me or someone else could continue later on.

**Comment by Kompwu** [comment]
@ai Hi, can you take a look ?

File: src/assets/SSSAsset.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+class SSSAsset extends CSSAsset {
+  constructor(name, options) {
+    options.css_parser = "sugarss";
+    super(name, options);
+    this.type = 'css';
+  }
+
+  parse(code) {
+    let root = postcss.parse(code, {from: this.name, to: this.name, syntax: sugarss});
```
Comment (ai) [inline src/assets/SSSAsset.js]: `syntax` => `parser`.

We need to set SugarSS only as input language (parser) and keep output language (stringifier) as default CSS

Comment (DoumanAsh) [reply src/assets/SSSAsset.js]: I think I also tried it initially, but for some reason postcss still parsed code as if it was CSS

Comment (ai) [reply src/assets/SSSAsset.js]: I will try to run tests after Def Con

Comment (Kompwu) [reply src/assets/SSSAsset.js]: I can try if you want.

Comment (ai) [reply src/assets/SSSAsset.js]: Yeap, it should be `parser:`.

Comment (Kompwu) [reply src/assets/SSSAsset.js]: So all that remains is to automate the configuration?

**Comment by Kompwu** [comment]
@DeMoorJasper Are you sure automating configuration is a good thing?
I don't think Parcel should do it.

**Comment by DeMoorJasper** [comment]
@Kompwu We are not automating config, we're adding a default?

I just thought it might be useful to actually have something with extension sugarSS be transpiled to css by default (it should still read the config and potentially even just pass along the sugarss file to postCSS without any transforms if it's configured). The rest of the PostCSS transforms will still apply as far as I know...

**Comment by Kompwu** [comment]
@DoumanAsh Can you finish?
All that remains is to add a default configuration if there is none.

**Comment by DoumanAsh** [comment]
@Kompwu To be honest I'm not sure why, but even when I place `parser`, it still fails in simple integration test.

The current commit also makes `sugarss` required dependency, would it be possible to make it optional?

**Comment by DeMoorJasper** [comment]
How does it add it as a dependency, you could use localRequire, that's how it's supposed to be used.

File: src/assets/SSSAsset.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+class SSSAsset extends CSSAsset {
+  constructor(name, options) {
+    options.css_parser = "sugarss";
+    super(name, options);
+    this.type = 'css';
+  }
+
+  parse(code) {
+    let root = postcss.parse(code, {from: this.name, to: this.name, parser: sugarss});
```
Comment (DeMoorJasper) [inline src/assets/SSSAsset.js]: let sugarss = await localRequire('sugarss')

File: src/assets/SSSAsset.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+const postcss = require('postcss');
+const sugarss = require('sugarss');
```
Comment (DeMoorJasper) [inline src/assets/SSSAsset.js]: should be a local require

File: src/assets/SSSAsset.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    this.type = 'css';
+  }
+
+  parse(code) {
+    let root = postcss.parse(code, {from: this.name, to: this.name, parser: sugarss});
+    return new CSSAst(code, root);
+  }
+}
+
+class CSSAst {
```
Comment (DeMoorJasper) [inline src/assets/SSSAsset.js]: Split this up into a seperate file, so both CSS and SugarSS Asset use it

File: src/transforms/postcss.js [hunk 5]
```diff
@@ -48,6 +48,15 @@ async function getConfig(asset) {
     config.plugins.push(postcssModules(postcssModulesConfig));
   }
 
+  if (config.parser === undefined && asset.options.css_parser !== undefined) {
+      const name = asset.options.css_parser;
+
+      if (!(typeof name === "string")) throw "Internal Error: Invalid PostCSS parser type. Should be string";
```
Comment (DeMoorJasper) [inline src/transforms/postcss.js]: `typeof name !== "string"`

File: src/transforms/postcss.js [hunk 6]
```diff
@@ -48,6 +48,15 @@ async function getConfig(asset) {
...
+
+      let parser = await localRequire(name, asset.name).catch(() => require(name));
```
Comment (DeMoorJasper) [inline src/transforms/postcss.js]: why does it catch with a require, that's what localRequire already does?

localRequire's should just throw if it fails, because require is just the same thing. (Unless you want to require something from inside parcel, but that should happen inside the transform not be a fallback as this could have unwanted side-effects)

File: src/assets/SSSAsset.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
+const postcss = require('postcss');
+const sugarss = require('sugarss');
+const CSSAsset = require('./CSSAsset.js');
+
+class SSSAsset extends CSSAsset {
+  constructor(name, options) {
+    options.css_parser = "sugarss";
```
Comment (DeMoorJasper) [inline src/assets/SSSAsset.js]: Why is this necessary?

Comment (DoumanAsh) [reply src/assets/SSSAsset.js]: I wanted to propagate to `transform` to automatically load parser, if it is necessary.

But not sure if it is good idea

**Comment by DoumanAsh** [comment]
Thanks for review, I applied majority of comments.

I'm not sure though how to automate configuration for PostCSS though.

My simple integration test is still failing me:

```
  sugarss
    1) should correctly parse SugarSS asset


  0 passing (478ms)
  1 failing

  1) sugarss
       should correctly parse SugarSS asset:
     CssSyntaxError: Missed semicolon
      at Input.error (node_modules\postcss\lib\input.js:119:22)
      at Parser.checkMissedSemicolon (node_modules\postcss\lib\parser.js:544:26)
      at Parser.decl (node_modules\postcss\lib\parser.js:269:50)
      at Parser.other (node_modules\postcss\lib\parser.js:168:18)
      at Parser.parse (node_modules\postcss\lib\parser.js:84:26)
      at Function.parse (node_modules\postcss\lib\parse.js:24:16)
      at SSSAsset.parse (src\assets\SSSAsset.js:16:24)
```

File: src/assets/Ast.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+class CSSAst {
```
Comment (DeMoorJasper) [inline src/assets/Ast.js]: Rename the file to CSSAst, as js and html also have an ast...

File: src/assets/Ast.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    if (this.dirty) {
+      this.css = '';
+      postcss.stringify(this.root, c => (this.css += c));
+    }
+
+    return this.css;
+  }
+}
+
+module.exports.CSSAst = CSSAst;
```
Comment (DeMoorJasper) [inline src/assets/Ast.js]: Just make this the default export `module.exports = CSSAst;`

File: src/assets/SSSAsset.js (hunk 7 above)
Comment (DeMoorJasper) [reply src/assets/SSSAsset.js]: You already do a localRequire in transform so not sure why this is necessary at all.

**Comment by DeMoorJasper** [comment]
You could check if the postCSS config already has a sugarss transform setup and skip this transform entirely if it has. But in the test it probably shouldn't have any postCSS config. Otherwise you're basically just testing postCSS

File: src/assets/CSSAsset.js [hunk 10]
```diff
@@ -3,6 +3,7 @@ const postcss = require('postcss');
 const valueParser = require('postcss-value-parser');
 const postcssTransform = require('../transforms/postcss');
 const CssSyntaxError = require('postcss/lib/css-syntax-error');
+const CSSAst = require('./Ast.js').CSSAst;
```
Comment (DeMoorJasper) [inline src/assets/CSSAsset.js]: Do the same thing for CSSAsset

**Comment by DeMoorJasper** [comment]
I've changed some stuff around and it should be working better now. You might want to cleanup and debug it a bit more as it's not actually working right now, and def not clean

File: src/assets/Ast.js (hunk 8 above)
Comment (DoumanAsh) [reply src/assets/Ast.js]: I was thinking that it would be better for future to have common source of asts?

Comment (DeMoorJasper) [reply src/assets/Ast.js]: In the future every asset will have it's own package so it will be seperate anyways

**Comment by DeMoorJasper** [comment]
Gonna open up a new PR that bases of off this one, with fixed code
